app-ng/src/handle-request.lisp

```diff
@@ -0,0 +1,103 @@
+
+(defun process-quil (quil)
+  "Prepare the PARSED-PROGRAM QUIL for more efficient execution. Currently this only includes remapping the qubits to a minimal sequential set from 0 to (num-qubits-used - 1). Return two values: the processed Quil code and the mapping vector.
+
+The mapping vector V specifies that the qubit as specified in the program V[i] has been mapped to qubit i."
+  (let* ((mapping (quil::compute-qubit-mapping quil))
+         (trivial-mapping-p
+           (loop :for x :across mapping
+                 :for i :from 0
+                 :always (= x i))))
+    (unless trivial-mapping-p
+      (format-log "Mapping qubits: ~{~A~^, ~}"
+                  (loop :for x :across mapping
+                        :for i :from 0
+                        :when (/= i x)
+                          :collect (format nil "~D -> ~D" x i)))
+      (quil::transform 'quil::compress-qubits quil))
+    (values quil mapping)))
+
+(defun get-random-state (arg)
+  (etypecase arg
+    (null (qvm:seeded-random-state nil))
+    (unsigned-byte (qvm:seeded-random-state arg))))
+
+(defun get-quil-instrs-field (hash-table)
+  (or (gethash "quil-instructions" hash-table)
+      (gethash "compiled-quil" hash-table)
+      (gethash "uncompiled-quil" hash-table)))
+
+(defun check-required-fields (hash-table &rest fields)
+  (dolist (field fields t)
+    (when (null (nth-value 1 (gethash field hash-table)))
+      (error "Expected the field ~A to exist." field))))
+
+(defun check-for-quil-instrs-field (hash-table)
+  (unless (get-quil-instrs-field hash-table)
+    (error "Expected a QUIL-INSTRUCTIONS type field to exist.")))
+
+(defun handle-post-request (request)
+  (when (null tbnl:*session*)
+    (tbnl:start-session))
+
+  (let* ((api-key (tbnl:header-in* ':X-API-KEY request))
+         (user-id (tbnl:header-in* ':X-USER-ID request))
+         (data (hunchentoot:raw-post-data :request request :force-text t))
+         (js (let* ((js (ignore-errors (yason:parse data))))
+               (unless (and (hash-table-p js)
+                            (check-required-fields js "type"))
+                 (error "Invalid request."))
+               js))
+         (type (gethash "type" js))
+         (gate-noise (gethash "gate-noise" js))
+         (measurement-noise (gethash "measurement-noise" js)))
+    (qvm:with-random-state ((get-random-state (gethash "rng-seed" js)))
+      (format-log "Got ~S request from API key/User ID: ~S / ~S" type api-key user-id)
+      (ecase (keywordify type)
+        ((:version)
+         (handle-version))
+
+        ((:delete-persistent-qvm)
+         (check-required-fields js "persistent-qvm-token")
+         (handle-delete-persistent-qvm (gethash "persistent-qvm-token" js)))
+
+        ((:make-persistent-qvm)
+         (check-required-fields js "simulation-method" "number-of-qubits")
+         (check-type (gethash "simulation-method" js) string)
+         (check-type (gethash "number-of-qubits" js) qvm::non-negative-fixnum)
```

Comment:
Note to self: switch to `alexandria:non-negative-fixnum` (here and anywhere else).

#160